STYLE_GUIDE.md

```diff
@@ -0,0 +1,164 @@
+This document is intended to provide more rigid structure for coding style and file structure.
+
+## References
+[BSG SystemVerilog Coding Guidelines](https://docs.google.com/document/d/1xA5XUzBtz_D6aSyIBQUwFk_kSUdckrfxa2uzGjMgmCU)
+
+[Freescale Verilog Guidelines](https://people.ece.cornell.edu/land/courses/ece5760/Verilog/FreescaleVerilog.pdf)
+
+[Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html)
+
+[lowRISCV Verilog Coding Style Guide](https://github.com/lowRISC/style-guides/blob/master/VerilogCodingStyle.md)
+
+## Verilog Style
+- No tabs in files, anywhere.  People get different results based on their editor.
+    - Exception: Makefiles. Tabs should be used to indent recipes, but not for general alignment
+- Align all things that have parallel structure (begin/end, case/endcase, commas, periods, etc.). This makes it easier to check the code and find the matching component.
+
+        if (x_p)  // notice how matching inputs are aligned
+          begin: foo
+            adder     (this_input, that_input, another_input);
+          end
+        else
+          begin: bar
+            wow_adder (this_input,          3, another_input);
+          end
+    
+    
+        if (foo)    // notice alignment of if/else, begin/end
+            x = 3;  // and optional alignment of x = statement.
+        else
+          begin
+            x = 1;
+	    	y = 0;
+          end
+
+- Use localparam keyword for derived parameters and input parameters that should not be set by the party instantiating the module.
+- Localparams in port lists are allowed, but only for sizing ports. Other generated localparams should go near their consumers.
+- Prefer input [1:0] to input logic [1:0] and output [1:0] to output logic [1:0]
+    - Ports with no type default to 'wire'. This forces you to use an 'assign' for outputs.  Since BSG style is to latch inputs and have combinatorial outputs, this style conveys hardware intent.
+- Ports and parameters should not have the module name as a prefix, since that is redundant. In some cases, the wires in the parent module may use the module name of the child to disambiguate. 
+
+        module foo 
+          #(parameter bar_p = "inv")
+           (input [bar_p-1:0] baz_i  // not foo_baz_i
+            );
+    
+        foo
+         #(.bar_p(foo_bar_p))
+         inst
+          (.baz_i(foo_baz_li));
+- Signals corresponding to a pipeline stage in a module which has multiple stages should have a pipeline stage suffix
+    - foobar_ex1_lo
+    - barbaz_tv_n
+- ‘0’, ‘1’, ‘2’, and ‘8’ (when used as byte width) are the only allowable magic numbers. Else, consider strongly the use of a localparam.
+- Prefer packages over \`includes. While \`defines are global and so this makes no difference to the LRM, BlackParrot style is to compile all packages first, so most tools will include definitions in the correct order.
```

Comment:
Grammar is a little funky here, revise.

Reply:
Revised, let me know if this makes more sense